simpa/core/device_digital_twins/illumination_geometries/ring_illumination.py

```diff
@@ -0,0 +1,122 @@
+        """
+        :param outer_radius_in_mm: The outer radius of the ring in mm.
+        :param inner_radius_in_mm: The inner radius of the ring in mm. If 0, should match the disk illumination.
+        :param lower_angular_bound: The lower angular bound in radians. If both bounds are 0, than no bound is applied.
+        Note that the bound of 0 starts from the x-axis on the right side and is applied clockwise.
+        :param upper_angular_bound: The upper angular bound in radians. If both bounds are 0, than no bound is applied.
+        Note that the bound is applied clockwise in relation to the lower bound.
+        :param device_position_mm: The device position in mm, the center of the ring.
+        If None, the position is defined as [0, 0, 0].
+        :param source_direction_vector: Direction of the illumination source.
+        If None, the direction is defined as [0, 0, 1].
+        :param field_of_view_extent_mm: Field of view which is defined as a numpy array of the shape \
+        [xs, xe, ys, ye, zs, ze], where x, y, and z denote the coordinate axes and s and e denote the start and end \
+        positions.
+        """
+        if device_position_mm is None:
+            device_position_mm = np.zeros(3)
+
+        if source_direction_vector is None:
+            source_direction_vector = np.array([0, 0, 1])
+
+        super(RingIlluminationGeometry, self).__init__(device_position_mm=device_position_mm,
+                                                       source_direction_vector=source_direction_vector,
+                                                       field_of_view_extent_mm=field_of_view_extent_mm)
+
+        assert inner_radius_in_mm >= 0, inner_radius_in_mm
+        assert outer_radius_in_mm >= inner_radius_in_mm, \
+            f"The outer radius ({outer_radius_in_mm}) has to be at least as large " \
+            f"as the inner radius ({inner_radius_in_mm})!"
+
+        assert lower_angular_bound >= 0, lower_angular_bound
+        assert upper_angular_bound >= lower_angular_bound, \
+            f"The outer radius ({upper_angular_bound}) has to be at least as large " \
+            f"as the inner radius ({lower_angular_bound})!"
+
+        self.outer_radius_in_mm = outer_radius_in_mm
+        self.inner_radius_in_mm = inner_radius_in_mm
+        self.lower_angular_bound = lower_angular_bound
+        self.upper_angular_bound = upper_angular_bound
+
+    def get_mcx_illuminator_definition(self, global_settings: Settings) -> dict:
+        """
+        Returns the illumination parameters for MCX simulations.
+        :param global_settings: The global settings.
+        :return: The illumination parameters as a dictionary.
+        """
+        assert isinstance(global_settings, Settings), type(global_settings)
+
+        source_type = Tags.ILLUMINATION_TYPE_RING
+
+        spacing = global_settings[Tags.SPACING_MM]
+
+        device_position = list(self.device_position_mm / spacing + 1)  # No need to round
```

Comment:
Add 0.5 insted of 1 because mcx puts the positions of source at the corner of a voxel and we want to have it in the center of the voxel

Reply:
From my experiments of visualizing the illuminated surface, I determined that using a offset of 0.5 leads to the right and bottom tissue edges being less illuminated, while using an offset of 1 results in a perfectly centered surface illumination. I think someone has to investigate the perfect offset for all of the illuminations (some illuminations apply rounding to the device position although it is not necessary)